test/e2e_node/gcloud/gcloud.go

```diff
@@ -0,0 +1,174 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gcloud
+
+import (
+	"errors"
+	"fmt"
+	"math/rand"
+	"os/exec"
+	//	"sync"
+
+	"net"
+	"net/http"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+)
+
+var freePortRegexp = regexp.MustCompile(".+:([0-9]+)")
+
+type TearDown func()
+
+type GCloudClient interface {
+	CopyAndWaitTillHealthy(sudo bool, remotePort string, timeout time.Duration, healthUrl string, bin string, args ...string) (*CmdHandle, error)
+}
+
+type gCloudClientImpl struct {
+	host string
+	zone string
+}
+
+type RunResult struct {
+	out []byte
+	err error
+	cmd string
+}
+
+type CmdHandle struct {
+	TearDown TearDown
+	Output   chan RunResult
+	LPort    string
+}
+
+func NewGCloudClient(host string, zone string) GCloudClient {
+	return &gCloudClientImpl{host, zone}
+}
+
+func (gc *gCloudClientImpl) Command(cmd string, args ...string) ([]byte, error) {
+	args = append([]string{"compute", "ssh", "--zone", gc.zone, gc.host, "--", cmd}, args...)
+	glog.Infof("Command %s", strings.Join(args, " "))
+	return exec.Command("gcloud", args...).CombinedOutput()
+}
+
+func (gc *gCloudClientImpl) TunnelCommand(sudo bool, lPort string, rPort string, cmd string, aa ...string) ([]byte, error) {
+	tunnelStr := fmt.Sprintf("-L %s:localhost:%s", lPort, rPort)
+	args := []string{"compute", "ssh", "--zone", gc.zone, "--ssh-flag", tunnelStr, gc.host, "--"}
+	if sudo {
+		args = append(args, "sudo")
+	}
+	args = append(args, cmd)
+	args = append(args, aa...)
+	glog.Infof("Command %s", strings.Join(args, " "))
+	return exec.Command("gcloud", args...).CombinedOutput()
+}
+
+func (gc *gCloudClientImpl) CopyToHost(from string, to string) ([]byte, error) {
+	rto := fmt.Sprintf("%s:%s", gc.host, to)
+	glog.Infof("Command gcloud compute copy-files --zone %s %s %s", gc.zone, from, rto)
+	return exec.Command("gcloud", "compute", "copy-files", "--zone", gc.zone, from, rto).CombinedOutput()
+}
+
+func (gc *gCloudClientImpl) CopyAndRunAsync(sudo bool, remotePort string, bin string, args ...string) *CmdHandle {
+	h := &CmdHandle{}
+	h.Output = make(chan RunResult)
+
+	rand.Seed(time.Now().UnixNano())
+
+	// Define where we will copy the temp binary
+	tDir := fmt.Sprintf("/tmp/gcloud-e2e-%d", rand.Int31())
```

Comment:
Ah, I see. And using them would create local directory instances just to get a name to use remote. It then would require cleaning up the local system as well as the remote. Likely more trouble than it's worth. Thanks for the clarification.